File: services/admin-service/app/services/health_checker.rb
class HealthChecker
  SERVICES = %w[auth-service file-service document-service collab-service notification-service search-service
                analytics-service audit-service].freeze

  ServiceStatus = Struct.new(:name, :status, :latency_ms, :message, keyword_init: true)

  def self.check_all
    services = SERVICES.map { |name| check_service(name) }
    overall = services.all? { |s| s.status == 'healthy' } ? 'healthy' : 'degraded'

    {
      status: overall,
      timestamp: Time.current.iso8601,
      services: services.map { |s| { name: s.name, status: s.status, latency_ms: s.latency_ms, message: s.message } },
      database: check_database,
      redis: check_redis
    }
  end

  def self.check_service(name)
    host = ENV.fetch("#{name.tr('-', '_').upcase}_HOST", name)
    port = ENV.fetch("#{name.tr('-', '_').upcase}_PORT", nil)
    url = "http://#{host}:#{port}/health" if port.present?

    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

    if url.present?
      uri = URI.parse(url)
      http = Net::HTTP.new(uri.host, uri.port)
      http.open_timeout = 2
      http.read_timeout = 2
      response = http.get(uri.path)
      latency = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000).round(1)

      ServiceStatus.new(name: name, status: response.code == '200' ? 'healthy' : 'unhealthy', latency_ms: latency)
    else
      ServiceStatus.new(name: name, status: 'unknown', latency_ms: 0, message: 'No endpoint configured')
    end
  rescue StandardError => e
    latency = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000).round(1)
    ServiceStatus.new(name: name, status: 'unhealthy', latency_ms: latency, message: e.message)
  end

  def self.check_database
    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    ActiveRecord::Base.connection.execute('SELECT 1')
    latency = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000).round(1)
    { status: 'healthy', latency_ms: latency }
  rescue StandardError => e
    { status: 'unhealthy', message: e.message }
  end

  def self.check_redis
    start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    redis_url = ENV.fetch('REDIS_URL', 'redis://localhost:6379/0')
    redis = Redis.new(url: redis_url, timeout: 2)
    redis.ping
    latency = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time) * 1000).round(1)
    { status: 'healthy', latency_ms: latency }
  rescue StandardError => e
    { status: 'unhealthy', message: e.message }
  end
end

File: services/admin-service/src/Services/HealthChecker.cs
using System.Diagnostics;
using Microsoft.EntityFrameworkCore;
using OtterWorks.AdminService.Data;
using OtterWorks.AdminService.Models.Dto;

namespace OtterWorks.AdminService.Services;

public interface IHealthChecker
{
    Task<ServiceHealthResponse> CheckAllAsync();
}

public class HealthChecker : IHealthChecker
{
    private static readonly string[] ServiceNames =
    [
        "auth-service", "file-service", "document-service", "collab-service",
        "notification-service", "search-service", "analytics-service", "audit-service"
    ];

    private readonly AdminDbContext _context;
    private readonly IConfiguration _configuration;
    private readonly IHttpClientFactory _httpClientFactory;

    public HealthChecker(AdminDbContext context, IConfiguration configuration, IHttpClientFactory httpClientFactory)
    {
        _context = context;
        _configuration = configuration;
        _httpClientFactory = httpClientFactory;
    }

    public async Task<ServiceHealthResponse> CheckAllAsync()
    {
        var services = new List<ServiceStatus>();
        foreach (var name in ServiceNames)
        {
            services.Add(await CheckServiceAsync(name));
        }

        var overall = services.All(s => s.Status == "healthy") ? "healthy" : "degraded";

        return new ServiceHealthResponse
        {
            Status = overall,
            Timestamp = DateTime.UtcNow.ToString("o"),
            Services = services,
            Database = await CheckDatabaseAsync(),
            Redis = CheckRedis(),
        };
    }

    private async Task<ServiceStatus> CheckServiceAsync(string name)
    {
        var envKey = name.Replace("-", "_", StringComparison.Ordinal).ToUpperInvariant();
        var host = _configuration[$"{envKey}_HOST"] ?? name;
        var port = _configuration[$"{envKey}_PORT"];

        if (string.IsNullOrEmpty(port))
        {
            return new ServiceStatus { Name = name, Status = "unknown", LatencyMs = 0, Message = "No endpoint configured" };
        }

        var sw = Stopwatch.StartNew();
        try
        {
            var client = _httpClientFactory.CreateClient();
            client.Timeout = TimeSpan.FromSeconds(2);
            var response = await client.GetAsync(new Uri($"http://{host}:{port}/health"));
            sw.Stop();
            return new ServiceStatus
            {
                Name = name,
                Status = response.IsSuccessStatusCode ? "healthy" : "unhealthy",
                LatencyMs = Math.Round(sw.Elapsed.TotalMilliseconds, 1),
            };
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new ServiceStatus
            {
                Name = name,
                Status = "unhealthy",
                LatencyMs = Math.Round(sw.Elapsed.TotalMilliseconds, 1),
                Message = ex.Message,
            };
        }
    }

    private async Task<ComponentHealth> CheckDatabaseAsync()
    {
        var sw = Stopwatch.StartNew();
        try
        {
            await _context.Database.ExecuteSqlRawAsync("SELECT 1");
            sw.Stop();
            return new ComponentHealth { Status = "healthy", LatencyMs = Math.Round(sw.Elapsed.TotalMilliseconds, 1) };
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new ComponentHealth { Status = "unhealthy", Message = ex.Message };
        }
    }

    private ComponentHealth CheckRedis()
    {
        var redisHost = _configuration["REDIS_HOST"] ?? "localhost";
        var redisPort = _configuration["REDIS_PORT"] ?? "6379";

        try
        {
            using var client = new System.Net.Sockets.TcpClient();
            var sw = Stopwatch.StartNew();
            client.Connect(redisHost, int.Parse(redisPort, System.Globalization.CultureInfo.InvariantCulture));
            sw.Stop();
            return new ComponentHealth { Status = "healthy", LatencyMs = Math.Round(sw.Elapsed.TotalMilliseconds, 1) };
        }
        catch (Exception ex)
        {
            return new ComponentHealth { Status = "unhealthy", Message = ex.Message };
        }
    }
}
